src/data/chats/model_loader.rs

```diff
@@ -1,88 +1,199 @@
+#[derive(Debug, Clone)]
+pub struct ModelLoaderError(Arc<anyhow::Error>);
+
+impl std::error::Error for ModelLoaderError {}
+
+impl std::fmt::Display for ModelLoaderError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ModelLoaderError: {}", self.0)
+    }
 }
 
-impl ModelLoader {
-    pub fn new(file: File) -> Self {
-        let (tx, rx) = channel();
-        Self {
-            complete: false,
-            file,
-            load_sender: tx,
-            load_receiver: rx,
-        }
+impl From<anyhow::Error> for ModelLoaderError {
+    fn from(err: anyhow::Error) -> Self {
+        Self(Arc::new(err))
     }
+}
 
-    pub fn load_model(&self, backend: &Backend) {
-        let (tx, rx) = channel();
-        let cmd = Command::LoadModel(
-            self.file.id.clone(),
-            LoadModelOptions {
-                prompt_template: None,
-                gpu_layers: moxin_protocol::protocol::GPULayers::Max,
-                use_mlock: false,
-                rope_freq_scale: 0.0,
-                rope_freq_base: 0.0,
-                context_overflow_policy:
-                    moxin_protocol::protocol::ContextOverflowPolicy::StopAtLimit,
-            },
-            tx,
-        );
+impl From<String> for ModelLoaderError {
+    fn from(err: String) -> Self {
+        Self(Arc::new(anyhow!(err)))
+    }
+}
 
-        let load_model_tx = self.load_sender.clone();
+impl From<&'static str> for ModelLoaderError {
+    fn from(err: &'static str) -> Self {
+        Self(Arc::new(anyhow!(err)))
+    }
+}
 
-        backend.command_sender.send(cmd).unwrap();
+/// All posible states in which the loader can be.
+#[derive(Debug, Default, Clone)]
+pub enum ModelLoaderStatus {
+    #[default]
+    Unloaded,
+    Loading,
+    Loaded,
+    Failed(#[allow(dead_code)] ModelLoaderError),
+}
 
-        thread::spawn(move || {
-            if let Ok(response) = rx.recv() {
-                match response {
-                    Ok(LoadModelResponse::Completed(_)) => {
-                        load_model_tx.send(Ok(())).unwrap();
-                    }
-                    Ok(_) => {
-                        eprintln!("Error loading model: Unexpected response");
-                        load_model_tx.send(
-                            Err(anyhow::anyhow!("Error loading model: Unexpected response"))
-                        ).unwrap();
-                    }
-                    Err(err) => {
-                        eprintln!("Error loading model: {:?}", err);
-                        load_model_tx.send(Err(err)).unwrap();
+#[derive(Default)]
+struct ModelLoaderInner {
+    status: ModelLoaderStatus,
+    file_id: Option<FileID>,
+}
+
+/// Unit for handling the non-blocking loading of models across threads.
+#[derive(Clone, Default)]
+pub struct ModelLoader(Arc<Mutex<ModelLoaderInner>>);
+
+impl ModelLoader {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn load_blocking(
```

Comment:
Having blocking functions by default instead of implicit threading functions makes code more composable and avoids creating extra threads when you are already in a separate synchronous thread (like the thread spawned by send message function).